File: 004/s.rb
#!/usr/bin/ruby

num = (ARGV[0] || '2').to_i

def makePairs(n)
  max = ('9' * n).to_i

  Enumerator.new do |yielder|
    xs = (1..max).reverse_each
    
    (0..(max - 1)).each do |d|
      ys = (1..(max - d)).reverse_each

      xs.zip(ys).each do |pair|
        yielder.yield pair if pair[0] && pair[1]
      end
    end
  end
end

def isPalindrome(s)
  (0..s.size/2).each do |i|
    return false if s[i] != s[s.size - 1 - i]
  end

  true
end

def findPalindrome(n)
  pairs = makePairs(n)
  pairs
    .zip(pairs.map { |x, y| x * y })
    .select { |pair, prod| isPalindrome(prod.to_s) }
    .max_by(&:last)
end

puts "N = #{num}"
puts "Solution: #{findPalindrome(num)}"

File: 004/s.hs
import Data.List (sortBy)

pairs :: Int -> [(Int, Int)]
pairs n = [ (x, y) | x <- [m, m-1 .. 1], y <- [m, m-1 .. 1]]
          where
            m = (read $ replicate n '9') :: Int

isPalindrome :: Int -> Bool
isPalindrome = let isPStr []     = True;
                   isPStr [_]    = True;
                   isPStr (x:xs) = (x == last xs) && isPStr (init xs)
               in isPStr . show

palindromes :: Int -> [(Int, Int, Int)]
palindromes n = sortBy (\(_,_,p1) (_,_,p2) -> compare p2 p1) [ (x, y, x * y) | (x, y) <- pairs n, isPalindrome (x * y)]

solution :: Int -> String
solution n = "Solution: " ++ show prod ++ " (" ++ show x ++ ", " ++ show y ++ ")"
             where
               (x, y, prod) = head . palindromes $ n

